lib/fileio/translators/translatorMayaReference.cpp

```diff
@@ -0,0 +1,489 @@
+
+    // Connect prim transform's message to reference's `associatedNode` attribute,
+    // so that the referenced nodes end up under the prim transform when loaded.
+    MFnReference refDependNode(referenceObject);
+    connectReferenceAssociatedNode(parentDag, refDependNode);
+
+  // Rename the reference node.  We want a unique name so that multiple copies
+  // of a given prim can each have their own reference edits. We make a name
+  // from the full path to the prim for which the reference is being created
+  // (and append "_RN" to the end, to indicate it's a reference node, just
+  // because).
+  //
+  MString uniqueRefNodeName = refNameFromPath(parentDag) + "_RN";
+  refDependNode.setName(uniqueRefNodeName);
+
+    // Now load the reference to properly trigger the kAfterReferenceLoad callback
+    MFileIO::loadReferenceByNode(referenceObject, &status);
+    CHECK_MSTATUS_AND_RETURN_IT(status);
+    {
+        // To avoid the error that USD complains about editing to same layer simultaneously from different threads,
+        // we record it as custom data instead of creating an attribute.
+        MString refDependNodeName = refDependNode.name();
+        VtValue value(std::string(refDependNodeName.asChar(), refDependNodeName.length()));
+        prim.SetCustomDataByKey(maya_associatedReferenceNode, value);
+    }
+
+    return MS::kSuccess;
+}
+
+MStatus 
+UsdMayaTranslatorMayaReference::UnloadMayaReference(const MObject& parent){
+    TF_DEBUG(PXRUSDMAYA_TRANSLATORS).Msg("MayaReferenceLogic::UnloadMayaReference\n");
+    MStatus status;
+    MFnDependencyNode fnParent(parent, &status);
+    if (status)
+    {
+        MPlug messagePlug = fnParent.findPlug("message", true, &status);
+        if (status)
+        {
+            MPlugArray referencePlugs;
+            messagePlug.connectedTo(referencePlugs, false, true);
+
+            // Unload the connected references.
+            for (uint32_t i = 0; i < referencePlugs.length(); ++i)
+            {
+                MObject temp = referencePlugs[i].node();
+                if (temp.hasFn(MFn::kReference))
+                {
+                    MFileIO::unloadReferenceByNode(temp, &status);
+                    CHECK_MSTATUS_AND_RETURN_IT(status);
+                }
+            }
+        }
+    }
+    return status;
+}
+
+MStatus 
+UsdMayaTranslatorMayaReference::connectReferenceAssociatedNode(MFnDagNode& dagNode, MFnReference& refNode)
+{
+    MPlug srcPlug = dagNode.findPlug("message", true);
+    /*
+       From the Maya docs:
+       > This message attribute is used to connect specific nodes that may be
+       > associated with this reference (i.e. group, locator, annotation). Use of
+       > this connection indicates that the associated nodes have the same
+       > lifespan as the reference, and will be deleted along with the reference
+       > if it is removed.
+     */
+    MStatus result;
+    MPlug destArrayPlug = refNode.findPlug("associatedNode", true);
```

Comment:
extract the MObject handle for the attribute upfront, to avoid the need for the binary search of string compares. e.g. upfront:

```c++
MNodeClass nc("reference");
UsdMayaTranslatorMayaReference::associatedNodeHandle = nc.attribute("associatedNode");
UsdMayaTranslatorMayaReference::messageHandle = nc.attribute("message");
```
Then later:
```c++
MPlug destArrayPlug(refNode.node(), associatedNodeHandle);
```